### ISSUE TITLE
module+ tries (and fails) to generate contracts for required definitions

### ISSUE THREAD

**Issue by jsmaniac** [racket/typed-racket#163]
We still have a small bug with submodules: when using a function from another typed module in the `(module+ test ...)`, typed/racket still tries to generate a contract for that function (and fails for the types for which a contract can't be generated).

Since the `module+` should use typed/racket, no contract should be generated.

Tested on 6.2.900.5 and 6.2.900.6 (snapshot 20150728-75e19c9).

```scheme
#lang typed/racket

;; Definition from the same module works in the (module+ test)
(: bar (case→ (→ 'a True) (→ 'b False)))
(define (bar x) (if (eq? x 'a) #t #f))

;; Definition from another module (nested like here or in another file) doesn't work in the (module+ test)
(module m typed/racket
  (: foo (case→ (→ 'a True) (→ 'b False)))
  (define (foo x) (if (eq? x 'a) #t #f))

  (provide foo))

(require 'm)

(module+ test
  ;; Okay
  (define b bar)
  ;; Type Checker: could not convert type to a contract;
  ;; function type has two cases of arity 1
  ;; identifier: foo
  ;; type: (case-> (-> a True) (-> b False))  in: foo
  (define f foo))
```

Using a `module*` works without problems:

```scheme
#lang typed/racket

;; Definition from the same module works in the (module* test typed/racket)
(: bar (case→ (→ 'a True) (→ 'b False)))
(define (bar x) (if (eq? x 'a) #t #f))

;; Definition from another module works in the (module* test typed/racket)
(module m typed/racket
  (: foo (case→ (→ 'a True) (→ 'b False)))
  (define (foo x) (if (eq? x 'a) #t #f))

  (provide foo))

(require 'm)

(provide bar foo)

(module* test typed/racket
  (require (submod ".."))
  (define b bar) ;; Okay
  (define f foo)) ;; Okay
```

**Comment by samth** [comment]
Here's a much simpler program with the same bug:

```
#lang racket

(module m typed/racket/base
  (define x 1)
  (provide x))

(module n typed/racket/base
  (require (submod ".." m))
  (module* a #f x))
```

**Comment by samth** [comment]
This bug is also the same as http://bugs.racket-lang.org/old/15118

**Comment by AlexKnauth** [comment]
Is this possibly related to https://github.com/racket/typed-racket/issues/130 ?

**Comment by takikawa** [comment]
Looks like this is due to the `typed-context?` flag giving the wrong value for that submodule. (because if you change the conditional inside `def-export` to unconditionally expand to the typed version, the problem seems to go away)

**Comment by samth** [comment]
Yes, I remember this problem. I think the behavior of module-begin in
submodules with #f as the language is different than I expect, and that
causes the problem. I don't think I came up with a solution the last time I
thought about it.

**Comment by takikawa** [comment]
Here's a minimized example that illustrates the expansion problem: http://pasterack.org/pastes/54064

**Comment by takikawa** [comment]
I think my previous example wasn't useful. I suspect http://pasterack.org/pastes/44433 is closer to what's going on in TR and has something to do with how modules are visited. (the last example where the `submod` require is moved into the `module*` matches up with what happens if you do the same in TR)

**Comment by samth** [comment]
Yes, this example corresponds to my recollection. I think this behavior
ought to change in Racket, but I don't know if that's possible.

**Comment by takikawa** [comment]
Closed by 241f04bcdb82d46af73890ccca484df3edbf27fa
